### PR TITLE
Gas Smashing

### PR DESCRIPTION
**Pull request by dariorussi**
## Description 
gas smashing per se is working and there are tests for it.
There are conversations in flight about how to do for the braking change with SDK and Wallet (@patrickkuo @666lcz @Jordan-Mysten).
Integration with Programmable Transaction (@tnowacki) should be good to go modulo some corrections.
@oxade please take a look if you can

## Test Plan 
Added unit tests

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [x] user-visible impact
- [x] breaking change for a client SDKs
- [x] breaking change for FNs (FN binary must upgrade)
- [x] breaking change for validators or node operators (must upgrade binaries)
- [x] breaking change for on-chain data layout
- [x] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-core/src/unit_tests/authority_tests.rs [hunk 1]
```diff
@@ -1916,36 +1916,6 @@ async fn test_type_argument_dependencies() {
     ));
 }
 
-#[tokio::test]
```
Comment (dariorussi) [inline crates/sui-core/src/unit_tests/authority_tests.rs:1919]: I am really not clear what this test was checking.
I have update the order in which input object checks and gas checks are performed and this test crashes (panic).
It was previously failing "gracefully" over a gas check (coin owned by sender).
However if I make the coin correctly owned by the sender, the test still crashes because it is truly sending a "transaction" that breaks invariants (only execution and no signing, where signing would have failed correctly).
So I have deleted it and hope someone can tell me what to do

File: crates/sui-adapter/src/execution_engine.rs [hunk 2]
```diff
@@ -155,7 +157,7 @@ fn execute_transaction<
         let execution_result = execution_loop::<Mode, _>(
             temporary_store,
             transaction_kind,
-            gas_object_id,
+            gas,
```
Comment (tnowacki) [inline crates/sui-adapter/src/execution_engine.rs:160]: It is not clear to me why this goes into the loop. 

Would it be possible to do something like
```
let gas_object_id = gas[0].0;
let gas_smash_changes = smash_gas(gas);
temporary_store.apply_object_changes(gas_smash_changes);
let result = charge_gas_for_object_read(temporary_store, &mut gas_status);
let mut result = result.and_then(|()| {
```
And then below
```
if execution_result.is_err() {
    temporary_store.apply_object_changes(gas_smash_changes);
    temporary_store.reset();
}
```

Long story short, programmable transactions do not work unless the gas is smashed before entering `execution_loop`. I also think doing it this will minimize the areas that need to change for execution?

Comment (dariorussi) [reply crates/sui-adapter/src/execution_engine.rs:160]: to be honest that is a bit the tragedy you have experienced as well.
That goes down only because Transfer uses it, one of the many special behavior of the gas flowing around the code base (waiting for ignition ;)).
I plan to readjust things when PT comes in and we clean up all the special behavior. I did not want to get intrusive.
If this creates a merge conflict I can adjust. 
Maybe I should study more how PT plans to use gas?

File: crates/sui-adapter/src/execution_engine.rs [hunk 3]
```diff
@@ -474,53 +484,6 @@ fn transfer_object<S>(
     Ok(())
 }
 
-fn check_coins(
```
Comment (dariorussi) [inline crates/sui-adapter/src/execution_engine.rs:477]: this moved in coin for better housing. Called from different place...

File: crates/sui-types/src/coin.rs [hunk 4]
```diff
@@ -161,6 +161,53 @@ impl Coin {
     }
 }
 
+pub fn check_coins(
```
Comment (dariorussi) [inline crates/sui-types/src/coin.rs:164]: this was moved from execution_engine.rs

File: sdk/typescript/test/e2e/txn-serializer.test.ts [hunk 5]
```diff
@@ -177,7 +177,7 @@ describe('Transaction Serialization and deserialization', () => {
       gasBudget: DEFAULT_GAS_BUDGET,
       gasPayment: coins[0].objectId,
     } as MoveCallTransaction;
-    const serArgsExpected = await new CallArgSerializer(
+    const setArgsExpected = await new CallArgSerializer(
```
Comment (666lcz) [inline sdk/typescript/test/e2e/txn-serializer.test.ts:180]:
```suggestion
    const serArgsExpected = await new CallArgSerializer(
```

This seems to be changed by the auto-formatter you are using

File: sdk/typescript/test/e2e/txn-serializer.test.ts [hunk 6]
```diff
@@ -198,10 +198,10 @@ describe('Transaction Serialization and deserialization', () => {
       gasBudget: DEFAULT_GAS_BUDGET,
       gasPayment: coins[0].objectId,
     } as MoveCallTransaction;
-    const serArgs = await new CallArgSerializer(
+    const setArgs = await new CallArgSerializer(
```
Comment (666lcz) [inline sdk/typescript/test/e2e/txn-serializer.test.ts:201]:
```suggestion
    const serArgs = await new CallArgSerializer(
```
ditto

File: sdk/typescript/src/signers/txn-data-serializers/local-txn-data-serializer.ts [hunk 7]
```diff
@@ -443,7 +443,7 @@ export class LocalTxnDataSerializer implements TxnDataSerializer {
           inputCoins: tx.Pay.coins.map((c) => c.objectId),
           recipients: tx.Pay.recipients,
           amounts: tx.Pay.amounts,
-          gasPayment: gasData.payment?.objectId,
+          gasPayment: gasData.payment?.[0].objectId,
```
Comment (666lcz) [inline sdk/typescript/src/signers/txn-data-serializers/local-txn-data-serializer.ts:446]: Can you add TODO comment here for supporting multiple gas objects

File: crates/sui-core/src/unit_tests/gas_tests.rs [hunk 8]
```diff
@@ -155,7 +155,7 @@ async fn test_native_transfer_gas_price_is_used() {
     // test overflow with insufficient gas
     let gas_balance = *MAX_GAS_BUDGET - 1;
     let gas_budget = *MAX_GAS_BUDGET;
-    let gas_price = 10;
+    let gas_price = 1;
```
[the remark this answers is not in the document]
Comment (dariorussi) [reply crates/sui-core/src/unit_tests/gas_tests.rs:158]: the way we change units the expected values few lines below would not match any longer.
So this was either changing here or the value expected below...

File: crates/sui-adapter/src/execution_engine.rs (hunk 2 above)
Comment (dariorussi) [reply crates/sui-adapter/src/execution_engine.rs]: I think I ended up doing what you mentioned here, take a look please

**Comment by oxade** [comment]
Whats the max number of gas objects that can be passed in? Otherwise we can do free work in gas smashing for low value coins.

File: crates/sui-types/src/messages.rs [hunk 9]
```diff
@@ -1296,18 +1288,56 @@ impl TransactionData {
...
+                budget: 0,
+            },
+            expiration: TransactionExpiration::None,
+        }
+    }
+
     pub fn new(
         kind: TransactionKind,
         sender: SuiAddress,
         gas_payment: ObjectRef,
```
Comment (tnowacki) [inline crates/sui-types/src/messages.rs:1318]: Thoughts about just making this a vector? 
Too many spots to change?

File: crates/sui-types/src/messages.rs [hunk 10]
```diff
@@ -1604,15 +1651,25 @@ impl TransactionData {
         let mut inputs = self.kind.input_objects()?;
 
         if !self.kind.is_system_tx() && !self.kind.is_pay_sui_tx() {
-            inputs.push(InputObjectKind::ImmOrOwnedMoveObject(
-                *self.gas_payment_object_ref(),
-            ));
+            inputs.extend(
+                self.gas()
+                    .iter()
+                    .map(|obj_ref| InputObjectKind::ImmOrOwnedMoveObject(*obj_ref)),
```
Comment (tnowacki) [inline crates/sui-types/src/messages.rs:1657]: If you want to be fancy
```suggestion
                    .iter()
                    .copied()
                    .map(InputObjectKind::ImmOrOwnedMoveObject),
```

File: crates/sui-core/src/unit_tests/authority_tests.rs [hunk 11]
```diff
@@ -4909,3 +4965,129 @@ fn test_choose_next_system_packages() {
...
+    }
+
+    // run an object creation transaction with the given amount of gas and coins, failure case
+    async fn run_and_check_err(reference_gas_used: u64, coin_num: u64, budget: u64) -> u64 {
+        let (sender, sender_key): (_, AccountKeyPair) = get_key_pair();
+        let gas_coins = make_gas_coins(sender, reference_gas_used, coin_num);
+        let gas_coin_ids: Vec<_> = gas_coins.iter().map(|obj| obj.id()).collect();
+        let (state, effects) = create_obj(sender, sender_key, gas_coins, budget).await;
+        // transaction is good
+        assert!(effects.status.is_err());
```
Comment (tnowacki) [inline crates/sui-core/src/unit_tests/authority_tests.rs:5054]: A bit lost. What is done to make this one a failure/error? Test looks _really_ similar to the one above

File: crates/sui-core/src/unit_tests/authority_tests.rs [hunk 12]
```diff
@@ -4909,3 +4965,129 @@ fn test_choose_next_system_packages() {
...
+    }
+
+    // run an object creation transaction with the given amount of gas and coins
+    async fn run_and_check_ok(reference_gas_used: u64, coin_num: u64, budget: u64) -> u64 {
+        let (sender, sender_key): (_, AccountKeyPair) = get_key_pair();
+        let gas_coins = make_gas_coins(sender, reference_gas_used, coin_num);
+        let gas_coin_ids: Vec<_> = gas_coins.iter().map(|obj| obj.id()).collect();
+        let (state, effects) = create_obj(sender, sender_key, gas_coins, budget).await;
+        // transaction is good
+        assert!(effects.status.is_ok());
```
Comment (tnowacki) [inline crates/sui-core/src/unit_tests/authority_tests.rs:5023]: Is the only difference between these two functions this one assert?

File: crates/sui-core/src/unit_tests/authority_tests.rs [hunk 13]
```diff
@@ -4909,3 +4965,129 @@ fn test_choose_next_system_packages() {
...
+    let reference_gas_used = gas_used + 1_000;
+    let three_coin_gas = run_and_check_ok(reference_gas_used, 3, reference_gas_used).await;
+    run_and_check_ok(reference_gas_used, 10, reference_gas_used - 100).await;
+
+    // make less then required to succeed
+    let reference_gas_used = gas_used - 10;
+    run_and_check_err(reference_gas_used, 2, reference_gas_used - 10).await;
+    run_and_check_err(reference_gas_used, 30, reference_gas_used).await;
+    // use a small amount less than what 3 coins above reported (with success)
+    run_and_check_err(three_coin_gas, 3, three_coin_gas - 1).await;
```
Comment (tnowacki) [inline crates/sui-core/src/unit_tests/authority_tests.rs:5092]: Should we have a test that checks that gas coins are always merged? Or is that what this is?

**Review by tnowacki — APPROVED** [review]

File: crates/sui-types/src/messages.rs (hunk 9 above)
Comment (dariorussi) [reply crates/sui-types/src/messages.rs:1318]: yep way toooooooo many, in time in time...

File: crates/sui-core/src/transaction_input_checker.rs [hunk 14]
```diff
@@ -21,26 +21,40 @@ use sui_types::{SUI_CLOCK_OBJECT_ID, SUI_CLOCK_OBJECT_SHARED_VERSION};
...
-    let gas_object_ref = transaction.gas_payment_object_ref();
-    let gas_object_refs = match tx_kind {
-        TransactionKind::Single(SingleTransactionKind::PaySui(p)) => p.coins.clone(),
-        TransactionKind::Single(SingleTransactionKind::PayAllSui(p)) => p.coins.clone(),
+
+    // Get the first coin (possibly the only one) and make it "the gas coin", then
+    // keep track of all others that can contribute to gas (gas smashing and special
+    // pay transactions).
+    let gas = transaction.gas();
+    let gas_object_ref = gas.get(0).unwrap();
```
[the remark this answers is not in the document]
Comment (dariorussi) [reply crates/sui-core/src/transaction_input_checker.rs:34]: I know I know, out of spite I may not even fix it....

File: crates/sui-core/src/unit_tests/authority_tests.rs (hunk 11 above)
Comment (dariorussi) [reply crates/sui-core/src/unit_tests/authority_tests.rs:5054]: I'll check

File: crates/sui-core/src/unit_tests/authority_tests.rs (hunk 12 above)
Comment (dariorussi) [reply crates/sui-core/src/unit_tests/authority_tests.rs:5023]: in my classic style I do not look for commonality in tests, but for differences, and I may fork more in coding. I'll check, thanks

File: crates/sui-core/src/unit_tests/authority_tests.rs (hunk 13 above)
Comment (dariorussi) [reply crates/sui-core/src/unit_tests/authority_tests.rs:5092]: I think they all do that or I hope so. But again the sooner we go in the more we can build test to make sure those edge conditions are proper
